src/Client/PacketReceiver.h

```diff
@@ -0,0 +1,143 @@
+#pragma once
+
+#if defined(OS_LINUX)
+
+#include <Client/IConnections.h>
+#include <Common/FiberStack.h>
+#include <Common/Fiber.h>
+
+namespace DB
+{
+
+/// Class for nonblocking packet receiving. It runs connection->receivePacket
+/// in fiber and sets special read callback which is called when
+/// reading from socket blocks. When read callback is called,
+/// socket and receive timeout are added in epoll and execution returns to the main program.
+/// So, you can poll this epoll file descriptor to determine when to resume
+/// packet receiving (beside polling epoll descriptor, you also need to check connection->hasPendingData(),
+/// because small packet can be read in buffer with the previous one, so new packet will be ready in buffer,
+/// but there is no data socket to poll).
```

Comment:
So, this means that an interface is not good enough :)
Again, I suppose that `resume` should return `variant(fd, packet, timeout)`.

Reply:
Also, now we have similar code in `PacketReceiver`, `ConnectionEstablisher` and `RemoteQueryExecutorReadContext`. It **does not** mean we should unite this code, but we should consider it.

Reply:
> So, this means that an interface is not good enough :)

I have an idea how to fix this problem with pending data, I will do this

Reply:
> I thought after finishing this PR we can discuss possible variants.

ok, that's fine